src/dryad.lisp

```diff
@@ -124,24 +130,39 @@ NOTE: In the basic implementation, these messages must be waiting for the DRYAD
 (define-process-upkeep ((dryad dryad) now) (SPROUTS-LOOP)
   "Loop over sprouted nodes, looking for ripe pairs."
   ;; if not everyone is sprouted, hold off
-  (unless (loop :for sprouted? :in (a:hash-table-values (dryad-sprouted? dryad))
-                :always sprouted?)
+  ;; NB: the loop returns T if the hash table is empty, so we additionally
+  ;;     guard against that
+  (unless (and (a:hash-table-values (dryad-sprouted? dryad))
+               (loop :for sprouted? :in (a:hash-table-values (dryad-sprouted? dryad))
+                     :always sprouted?))
     (process-continuation dryad `(SPROUTS-LOOP))
     (finish-with-scheduling))
   (let ((addresses (a:hash-table-keys (dryad-sprouted? dryad))))
     (flet ((payload-constructor ()
              (make-message-values :reply-channel (register)
-                                  :values '(match-edge))))
+                                  :values '(match-edge pistil
+                                            parent children))))
       (with-replies (replies) (send-message-batch #'payload-constructor addresses)
-        ;; make sure everyone has a match. any that doesn't is in a blossom
-        ;; which needs to be expanded.
-        (loop :for address :in addresses
-              :for reply :in replies
-              :when (null (first reply))
-                :do (process-continuation dryad
-                                          `(SEND-EXPAND ,address)
-                                          `(SPROUTS-LOOP))
-                    (finish-with-scheduling))
+        ;; verify some local state
+        (let (mid-augment?)
+          (loop :for address :in addresses
+                :for (match-edge pistil parent children) :in replies
+                ;; if we have a parent or children, we're in the middle of an augment
+                :when (or parent children)
+                  :do (setf mid-augment? t)
+                      ;; if we don't have a match, then we are inside a macrovertex that
+                      ;; needs to be expanded
+                :when (null match-edge)
+                  :do (assert pistil () "How are we matchless but have no pistil?")
+                      (process-continuation dryad
+                                            `(SEND-EXPAND ,address)
+                                            `(SPROUTS-LOOP))
+                      (finish-with-scheduling))
+          ;; if we're in the middle of an augment, we should pause for a bit
+          (when mid-augment?
+            (process-continuation dryad
+                                  `(SPROUTS-LOOP))
+            (finish-with-scheduling)))
```

Comment:
It's deliberate that you defer this to after the loop, so that you might prefer to SEND-EXPAND to something in the middle of augmenting?

Reply:
yeah